sound/soc/sof/ipc.c

```diff
@@ -612,93 +612,97 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
 }
 EXPORT_SYMBOL(snd_sof_ipc_stream_posn);
 
-/*
- * IPC get()/set() for kcontrols.
- */
+static int sof_get_ctrl_ipc_dir(enum sof_ipc_ctrl_type ctrl_type)
+{
+	int send;
+
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_DATA_GET:
+		send = 0;
+		break;
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		send = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
 
-int snd_sof_ipc_set_comp_data(struct snd_sof_ipc *ipc,
-			      struct snd_sof_control *scontrol, u32 ipc_cmd,
-			      enum sof_ipc_ctrl_type ctrl_type,
-			      enum sof_ipc_ctrl_cmd ctrl_cmd)
+	return send;
+}
+
+static void sof_set_get_mmap(struct snd_sof_dev *sdev,
+			     struct snd_sof_control *scontrol,
+			     struct sof_ipc_ctrl_data *cdata,
+			     int send)
 {
-	struct snd_sof_dev *sdev = ipc->sdev;
-	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	int err;
+	u32 send_bytes;
 
-	/* read firmware volume */
-	if (scontrol->readback_offset != 0) {
-		/* we can read value header via mmaped region */
-		snd_sof_dsp_block_write(sdev, sdev->mmio_bar,
-					scontrol->readback_offset, cdata->chanv,
-					sizeof(struct sof_ipc_ctrl_value_chan) *
-					cdata->num_elems);
+	/* write/read value header via mmaped region */
+	send_bytes = sizeof(struct sof_ipc_ctrl_value_chan) *
+		cdata->num_elems;
 
-	} else {
-		/* write value via slower IPC */
-		cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
-		cdata->cmd = ctrl_cmd;
-		cdata->type = ctrl_type;
-		cdata->rhdr.hdr.size = scontrol->size;
-		cdata->comp_id = scontrol->comp_id;
-		cdata->num_elems = scontrol->num_channels;
-
-		/* send IPC to the DSP */
-		err = sof_ipc_tx_message(sdev->ipc,
-					 cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size,
-					 cdata, cdata->rhdr.hdr.size);
-		if (err < 0) {
-			dev_err(sdev->dev, "error: failed to set control %d values\n",
-				cdata->comp_id);
-			return err;
-		}
-	}
+	if (send)
+		snd_sof_dsp_block_write(sdev, sdev->mmio_bar,
+					scontrol->readback_offset,
+					cdata->chanv, send_bytes);
 
-	return 0;
+	else
+		snd_sof_dsp_block_read(sdev, sdev->mmio_bar,
+				       scontrol->readback_offset,
+				       cdata->chanv, send_bytes);
```

Comment:
you are doing two changes at once, it'd be easier to read if you had a first patch with the get_set changes, then a second one where the block_read and get_mmap changes are introduced.

Reply:
ok, will refactor

Reply:
@plbossart updated, I tried to chop it to pieces, hope it is more readable now.